File: 0002_Add_Two_Numbers/Add_Two_Numbers.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
  result = []
  temp = 0
  while l1 && l2
    result << (l1.val + l2.val + temp) % 10
    temp = (l1.val + l2.val + temp) / 10
    l1 = l1.next
    l2 = l2.next
  end
  while l1
    result << (l1.val + temp) % 10
    temp = (l1.val + temp) / 10
    l1 = l1.next
  end
  while l2
    result << (l2.val + temp) % 10
    temp = (l2.val + temp) / 10
    l2 = l2.next
  end
  result.push temp if temp.nonzero?
  result
end

File: 0002_Add_Two_Numbers/Add_Two_Numbers.py
# -*- coding: utf-8 -*-

"""
You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Example:

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)

Output: 7 -> 0 -> 8

Explanation: 342 + 465 = 807.
"""


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        result = ListNode(0)
        current = result
        temp = 0
        while l1 or l2:
            v1 = l1.val if l1 else 0
            v2 = l2.val if l2 else 0
            current.next = ListNode((v1 + v2 + temp) % 10)
            current = current.next
            temp = (v1 + v2 + temp) // 10
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        current.next = ListNode(1) if temp else None
        return result.next

